src/cards/polymorph.ts

```diff
@@ -0,0 +1,165 @@
+import { playDefaultSpellSFX } from './cardUtils';
+import { Spell, addTarget, allModifiers, getCurrentTargets, refundLastSpell } from './index';
+import { CardRarity, probabilityMap } from '../types/commonTypes';
+import * as Unit from '../entity/Unit';
+import * as Pickup from '../entity/Pickup';
+import { allUnits } from '../entity/units';
+import { isModActive } from '../registerMod';
+import Underworld from '../Underworld';
+import { IPickup, isPickup, pickups } from '../entity/Pickup';
+import { bossmasonUnitId } from '../entity/units/deathmason';
+import floatingText from '../graphics/FloatingText';
+import seedrandom from 'seedrandom';
+import { chooseObjectWithProbability, chooseOneOf, chooseOneOfSeeded } from '../jmath/rand';
+
+export const polymorphId = 'Polymorph';
+const spell: Spell = {
+  card: {
+    id: polymorphId,
+    category: CardCategory.Curses,
+    supportQuantity: true,
+    manaCost: 40,
+    healthCost: 0,
+    expenseScaling: 2,
+    probability: probabilityMap[CardRarity.RARE],
+    thumbnail: 'spellIconPolymorph.png',
+    sfx: 'purify',
+    description: ['spell_polymorph'],
+    effect: async (state, card, quantity, underworld, prediction) => {
+      const targets = getCurrentTargets(state);
+      if (targets.length == 0) {
+        refundLastSpell(state, prediction, 'No valid targets. Cost refunded.');
+        return state;
+      }
+
+      playDefaultSpellSFX(card, prediction);
+
+      for (let i = 0; i < quantity; i++) {
+        for (const target of targets) {
+          if (Unit.isUnit(target)) {
+            const unit = target as Unit.IUnit;
+            const unitSource = allUnits[unit.unitSourceId]
+
+            let possibleUnitTypes = Object.values(allUnits).filter(u => isModActive(u, underworld) && u.id != unitSource?.id);
+            if (unit.isBoss) {
+              possibleUnitTypes = possibleUnitTypes
+                .filter(u => u.unitProps.isBoss);
+            } else {
+              // Get all units with budget and spawn chance
+              possibleUnitTypes = possibleUnitTypes
+                .filter(u => u.spawnParams && u.spawnParams.probability > 0 && u.spawnParams.budgetCost);
+            }
+
+            // We have to seed this to prevent multiplayer desync
+            const seed = seedrandom(`${underworld.turn_number} -${target.id} `);
+            const chosenIndex = chooseObjectWithProbability(possibleUnitTypes.map((p, index) => {
+              // Units are weighted by their difference in budget.
+              // Units twice as far away in the budget are half as likely to be chosen.
+              const budgetDiff = Math.abs((unitSource?.spawnParams?.budgetCost || 0) - (p.spawnParams?.budgetCost || 0));
+              return { index, probability: (budgetDiff == 0) ? 1 : 1 / budgetDiff }
+            }), seed)?.index;
+
+            const chosenUnitType = (chosenIndex != undefined) ? possibleUnitTypes[chosenIndex] : undefined;
```

Comment:
No need to use .index on line 61 and then use the index on line 63.  Just have chooseObjectWithProbability return `p` and as unit {unit:p, probability: etc} instead of {index, probability} and then do .unit on line 61